Fix password displayed as username in Firefox password manager dialog

danxuliu opened the pull request:
Fixes (in master) #7031

The problem was (besides the [password manager ignoring `autocomplete="off"`](https://developer.mozilla.org/en-US/docs/Web/Security/Securing_your_site/Turning_off_form_autocompletion#The_autocomplete_attribute_and_login_fields)) that the password input element (in the HTML form) is cloned in a text input element, which is used to show the password in plain text when clicking on the _Show password_ button. As it was a text input immediately followed by a password input Firefox seemed to assume that it had to be the username and ignored the real username field, no matter the value set for the `autocomplete` attribute (`autocomplete="off"` for the whole form, `autocomplete="street-name"` (because it is a [supported `autocomplete` value](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#attr-autocomplete) that has nothing to do with login) or `autocomplete="new-password"` for the cloned text input, `autocomplete="name/username/email"` for the username text input, nor combinations of those).

Now the cloned text input is added after the password input, so Firefox no longer thinks that the cloned text input is the username field and the password manager dialog shows the proper username instead.

Chromium was not affected by the original issue (it also offers to save the credentials, but its password manager dialog showed the proper username), and it still works as before after the change.

danxuliu committed:
Fix password displayed as username in Firefox password manager dialog

When finishing the setup of Nextcloud through the WebUI (setting admin
user and database) Firefox offers to save the username and password.
However, the password was shown in both the username and password fields
of the Firefox password manager dialog.

The problem was that the password input element (in the HTML form) is
cloned in a text input element, which is used to show the password in
plain text when clicking on the "Show password" button. As it was a text
input immediately followed by a password input Firefox seemed to assume
that it had to be the username and ignored the real username field, no
matter the value set for the "autocomplete" attribute. Now the cloned
text input is added after the password input, so Firefox no longer
thinks that the cloned text input is the username field and the password
manager dialog shows the proper username instead.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

rullzer reviewed (approved):
Sure :)

MorrisJobke commented:
* [ ] the password strength meter on the personal settings page now moves above the input field:

![bildschirmfoto 2018-01-05 um 10 47 36](https://user-images.githubusercontent.com/245432/34603885-0a3a2078-f206-11e7-8308-89f09f1110fe.png)
![bildschirmfoto 2018-01-05 um 10 47 41](https://user-images.githubusercontent.com/245432/34603886-0a59c81a-f206-11e7-95fc-f306444493f0.png)

danxuliu commented:
>  the password strength meter on the personal settings page now moves above the input field

Thanks, I had not noticed it. The problem is that _strengthify_ also [adds the password strength meter after the element it is applied on](https://github.com/nextcloud/server/blob/afcb41886ecaf3b05a21f4379f6386fba5ff7954/core/vendor/strengthify/jquery.strengthify.js#L179) and as [it is applied on `#pass2`](https://github.com/nextcloud/server/blob/4c3eb80a9102ceaf08d5143eaba8877a8a43e57d/settings/js/settings/personalInfo.js#L388) after [calling showPassword](https://github.com/nextcloud/server/blob/4c3eb80a9102ceaf08d5143eaba8877a8a43e57d/settings/js/settings/personalInfo.js#L144) it ends between the password input and the plain text input clone.

As the text input clone is updated when the password input is updated and vice versa I will check if it is possible to apply _strengthify_ to the text input clone instead.

danxuliu committed:
Add strengthify bar after plain text input clone of password input

To show the password in plain text "showPassword" adds a text input
after the password input and swaps their visibility depending on whether
the password has to be shown in plain text or not. In a similar way,
"strengthify" by default adds the strength bar after the input element
it was called on. Due to this, if "showPassword" is called before
"strengthify" on the same password input then the strength bar ends
between the password input and the text input, and when the text input
is shown it appears below the strength bar.

To fix this now the strength bar is added after the text input in those
places in which "strengthify" was called after "showPassword" on the
same element.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu commented:
>  I will check if it is possible to apply strengthify to the text input clone instead.

No, that approach is not valid due to how the events are triggered in the password input and the text input clone.

The strength bar problem can be fixed (I have tested it), besides using absolute CSS positions (which I would prefer to avoid), by making possible to specify the element that the strength bar should be added after.

@MorrisJobke I guess that I should send a pull request with the needed change in https://github.com/MorrisJobke/strengthify, and then just update the version in [bower.json](https://github.com/nextcloud/server/blob/6d2bff46c95615a774e9701c02da0cd4fc3561b5/bower.json#L28) and the code in _core/vendor_, right? Am I missing anything?

MorrisJobke commented:
That’s right - maybe we should not use it as external dependency and just maintain it in here as I don’t maintain it anyways - shame on me :/

rullzer commented:
So what is the status here? Fix soonish? Move to 14?

danxuliu commented:
@rullzer
>  So what is the status here? Fix soonish? Move to 14?

It is almost done; needs https://github.com/MorrisJobke/strengthify/pull/20 and using the new parameter when calling `strengthify` here.

MorrisJobke commented:
> It is almost done; needs MorrisJobke/strengthify#20 and using the new parameter when calling strengthify here.

Just released it as 0.5.3: https://github.com/MorrisJobke/strengthify/releases/tag/0.5.3

rullzer committed:
Bump strengthify to 0.5.3

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
Bumd the js dep... but it already looks ok now in chromium for me...

rullzer commented:
ah no caching :P

danxuliu commented:
Thanks for the library bump :-)

I have pushed the commit that uses the new `$addAfter` and [fixes the strength bar changing its position when the password is shown](https://github.com/nextcloud/server/pull/7709#issuecomment-355514262).

It should fix it too in the _Renew password_ of LDAP... although **I have not tested** it myself in that case, just updated the code.

No changes were needed in the installation page, as in that case `showPassword` was called after `strengthify`.

rullzer reviewed (approved):
whoohoo! still works!

MorrisJobke commented on apps/user_ldap/js/renewPassword.js:
```diff
@@ -46,5 +46,6 @@ $(document).ready(function() {
 			t('core', 'Strong password')
 		],
 		drawTitles: true,
+		$addAfter: $('input[name="newPassword-clone"'),
```
There is a missing `]` which is also logged in the JS console. Let me fix this.

MorrisJobke commented on settings/js/settings/personalInfo.js:
```diff
@@ -395,6 +395,7 @@ $(document).ready(function () {
 			t('settings', 'Strong password')
 		],
 		drawTitles: true,
+		$addAfter: $('input[name="newpassword-clone"'),
```
Same here 🙈

MorrisJobke replied:
<img width="540" alt="bildschirmfoto 2018-01-11 um 14 07 27" src="https://user-images.githubusercontent.com/245432/34826932-c7f20064-f6d8-11e7-90e5-f52ecea02461.png">

MorrisJobke reviewed (approved):
Tested and works 👍

danxuliu commented:
Backport is in #7797
